### ISSUE TITLE
Register<Func<Foo, IBar>>(c) vs Register<Foo, IBar>(c, foo)

### ISSUE THREAD

**Issue by 0shi** [seesharper/LightInject#485]
Gidday,
I was running into an issue yesterday as follows -
- We have a class IBar implemented by Bar which uses a Foo in its constructor
- We have a number of Classes which are scoped to a single Foo. In these classes, we resolve a single instance of IBar which we use for the lifetime of the Foo
- We have a number of classes which are scoped to potentially multiple Foos over their lifetime. In these classes, we resolve a factory as Func<Foo, IBar> which takes in a Foo and returns an instance of IBar

I was making use of the following registrations and encountering a StackOverFlowException:
`            
serviceContainer.Register<Foo, IBar>((c, tenant) => { return Barfactory(c, tenant); });
serviceContainer.Register<IBar>((c) => { var foo = c.GetInstance<Foo>(); return c.GetInstance<Foo, IBar>(foo); });
serviceContainer.GetInstance<IBar>();
`

I changed this as follows, and am no longer encountering the StackOverFlowException:
`            
            serviceContainer.Register<Func<Foo, IBar>>(c =>
            {
                return (foo) => Barfactory(c, foo);
            }, new PerContainerLifetime()); // This PerContainerLifeTime is for the Function, not the result of the Function
            serviceContainer.Register<IBar>((c) =>
            {
                var foo = c.GetInstance<Foo>();
                var factory = c.GetInstance<Func<Foo, IBar>>();
                return factory(foo);
            });
serviceContainer.GetInstance<IBar>();
`

**Comment by seesharper** [comment]
yeah, a repro project would be great. Just put it up here on GitHub so we can have a look 👍

**Comment by 0shi** [comment]
No worries, I put this together earlier today, just pushed it now: [Repro project (.net 461, LI 5.4)](https://github.com/0shi/LightInjectStackOverflowRepro)
